lint/auth_account_parameter_analyzer.go

```diff
@@ -0,0 +1,86 @@
+/*
+ * Cadence-lint - The Cadence linter
+ *
+ * Copyright 2019-2022 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lint
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/tools/analysis"
+)
+
+var AuthAccountParameterAnalyzer = (func() *analysis.Analyzer {
+
+	elementFilter := []ast.Element{
+		(*ast.FunctionDeclaration)(nil),
+		(*ast.SpecialFunctionDeclaration)(nil),
+	}
+
+	return &analysis.Analyzer{
+		Description: "Detects functions with AuthAccount type parameters",
+		Requires: []*analysis.Analyzer{
+			analysis.InspectorAnalyzer,
+		},
+		Run: func(pass *analysis.Pass) interface{} {
+			inspector := pass.ResultOf[analysis.InspectorAnalyzer].(*ast.Inspector)
+
+			location := pass.Program.Location
+			report := pass.Report
+
+			inspector.Preorder(
+				elementFilter,
+				func(element ast.Element) {
+					functionDeclaration, isFD := element.(*ast.FunctionDeclaration)
+					specialFunctionDeclaration, isSFD := element.(*ast.SpecialFunctionDeclaration)
+					var parameterList *ast.ParameterList
+					if isSFD && specialFunctionDeclaration.DeclarationKind() == common.DeclarationKindInitializer {
+						parameterList = specialFunctionDeclaration.FunctionDeclaration.ParameterList
+					} else if isFD {
+						parameterList = functionDeclaration.ParameterList
+					} else {
+						return
+					}
```

Comment:
maybe just use a switch statement here? i.e. `switch declaration := element.(type)`

Reply:
Neat! Updated in https://github.com/onflow/cadence-tools/pull/85/commits/ec125ba0fe327fed41f5f9b6a436a6dd8b17d587